wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -0,0 +1,487 @@
+import com.squareup.kotlinpoet.KModifier.OVERRIDE
+import com.squareup.kotlinpoet.KModifier.PRIVATE
+import com.squareup.kotlinpoet.LONG
+import com.squareup.kotlinpoet.NameAllocator
+import com.squareup.kotlinpoet.ParameterSpec
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.asClassName
+import com.squareup.wire.EnumAdapter
+import com.squareup.wire.FieldEncoding
+import com.squareup.wire.ProtoAdapter
+import com.squareup.wire.ProtoReader
+import com.squareup.wire.ProtoWriter
+import com.squareup.wire.WireEnum
+import com.squareup.wire.schema.EnumType
+import com.squareup.wire.schema.Field
+import com.squareup.wire.schema.MessageType
+import com.squareup.wire.schema.ProtoFile
+import com.squareup.wire.schema.ProtoType
+import com.squareup.wire.schema.Schema
+import com.squareup.wire.schema.Type
+import okio.ByteString
+import java.util.*
+
+class KotlinGenerator private constructor(
+    val schema: Schema,
+    private val nameToKotlinName: Map<ProtoType, ClassName>,
+    private val emitAndroid: Boolean
+) {
+  /** Returns the Kotlin type for [protoType]. */
+  fun typeName(protoType: ProtoType) = nameToKotlinName.getValue(protoType)
+
+  /** Returns the full name of the class generated for `type`.  */
+  fun generatedTypeName(type: Type) = typeName(type.type())
+
+  fun generateType(type: Type): TypeSpec = when (type) {
+    is MessageType -> generateMessage(type)
+    is EnumType -> generateEnum(type)
+    else -> error("Unknown type $type")
+  }
+
+  // TODO Make this a CACHE
+  private fun nameAllocator(message: Type): NameAllocator {
+    val allocator = NameAllocator()
+    when (message) {
+      is EnumType -> {
+        allocator.newName("value", "value")
+        allocator.newName("ADAPTER", "ADAPTER")
+        message.constants().forEach { constant ->
+          allocator.newName(constant.name(), constant)
+        }
+      }
+      is MessageType -> {
+        allocator.newName("unknownFields", "unknownFields")
+        allocator.newName("ADAPTER", "ADAPTER")
+        allocator.newName("reader", "reader")
+        if (emitAndroid) {
+          allocator.newName("CREATOR", "CREATOR")
+        }
+        message.fields().forEach { field ->
+          allocator.newName(field.name(), field)
+        }
+      }
+    }
+    return allocator
+  }
+
+  private fun generateMessage(type: MessageType): TypeSpec {
+    val className = typeName(type.type())
+
+    val classBuilder = TypeSpec.classBuilder(className)
+        .addModifiers(DATA)
+        .addType(generateAdapter(type))
+
+    if (emitAndroid) {
+      addAndroidCodeToMessage(classBuilder, type)
+    }
+
+    addMessageConstructor(type, classBuilder)
+
+    type.nestedTypes().forEach { classBuilder.addType(generateType(it)) }
+
+    return classBuilder.build()
+  }
+
+  /**
+   * Example
```

Comment:
Wanna make this proper KDoc? (Just wrap the code bits in three backticks above and below)

Reply:
done